### PR TITLE
Allow editing non-existent configs

### PR DESCRIPTION
**Pull request by qfel**
Addresses #1571.

`jj config edit --user` and `jj config set --user` will now pick a default config location if no existing is found.

Platform-specific config directory is preferred over the home directory.

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -190,30 +190,83 @@ impl LayeredConfigs {
...
-                xdg_config_path.clone(),
-                home_config_path.clone(),
-            )),
-            _ => Ok(platform_specific_config_path.or(home_config_path)),
-        }
+        return Ok(ConfigPath::Existing(PathBuf::from(config_path)));
+    }
+    // TODO: Should we drop the final `/config.toml` and read all files in the
+    // directory?
+    let xdg_config_path = ConfigPath::from(dirs::config_dir().map(|mut config_dir| {
```
Comment (necauqua) [inline src/config.rs:234]: nit: it's only xdg on linux, original name was long but more correct.

File: src/config.rs [hunk 2]
```diff
@@ -190,30 +190,83 @@ impl LayeredConfigs {
...
+    }));
+    use ConfigPath::*;
+    match (xdg_config_path, home_config_path) {
+        (Existing(xdg_config_path), Existing(home_config_path)) => Err(
+            ConfigError::AmbiguousSource(xdg_config_path, home_config_path),
+        ),
+        (Existing(path), _) | (_, Existing(path)) => Ok(Existing(path)),
+        // Order matters: prefer creating new configs in the platform-specific directory instead of
+        // the home directory.
+        (New(path), _) | (_, New(path)) => Ok(New(path)),
```
Comment (necauqua) [inline src/config.rs:251]: This does not work because it's unlikely that a `jj` folder exists in the platform dir, which is checked in the `ConfigPath::new`

**Review by necauqua — CHANGES_REQUESTED** [review]

File: src/config.rs (hunk 2 above)
Comment (qfel) [reply src/config.rs:251]: True, but I'm not sure if that means we should do something else.

If the folder exists, creating the config there seems like the right thing. If it doesn't, creating the jj subfolder (providing the platform-specific dir exists) seems questionable. Most tools don't behave this way, and while I could see `jj config set` doing that, `edit` would be more of an issue - we'd have to create the dir before opening the editor, which would leave it empty if the user changed their mind.

Basically I don't think it makes much sense to remove the first pattern and relax (Unavailable, Unavailable) into (_, Unavailable) just because it's very unlikely it's going to make a difference - and I'm not sure there's much of an alternative approach.

Should we just create the jj dir?

Comment (yuja) [reply src/config.rs:251]: > Should we just create the jj dir?

If we prefer platform-specific dir (and if the base dir exists), yes, the `jj` subdirectory should be created. 
I think `jj config edit` can also write a sample config file (with comment to the documentation.)

Comment (necauqua) [reply src/config.rs:251]: The config file in home dir is `.jjconfig.toml` (ew), the config file in platform-specific dir is `jj/config.toml`, so I see no contradiction in creating the `jj` folder, it's just that the current logic doesn't work like that.

For "edit" I would expect it to write a default config there and then edit that, yes

Comment (qfel) [reply src/config.rs]: OK then, made it create the needed directories and files. Initial config template, if any, can be tackled in a separate PR.

Also changed the code quite a bit to add unit tests.

File: src/config.rs [hunk 3]
```diff
@@ -190,28 +192,126 @@ impl LayeredConfigs {
...
+            // TODO: We should probably support colon-separated (std::env::split_paths)
+            // paths here
+            let path = PathBuf::from(path);
+            if allow_create {
+                if let Err(e) = create_config_file(&path) {
+                    if e.kind() != std::io::ErrorKind::AlreadyExists {
+                        return Err(e.into());
+                    }
+                }
+            }
```
Comment (yuja) [inline src/config.rs:255]: Perhaps, `config_path()` can be a pure function that returns `Result<ConfigPath, _>` enum, and any side effect like creating empty config file can be moved to `new_config_path()`.

File: src/config.rs [hunk 4]
```diff
@@ -632,4 +732,286 @@ mod tests {
...
+                    std::fs::create_dir_all(parent)?;
+                }
+                std::fs::File::create(path)?;
+            }
+            Ok(())
+        }
+
+        fn run(self) -> anyhow::Result<()> {
+            use anyhow::anyhow;
+            let tmp = tempfile::tempdir()?;
```
Comment (yuja) [inline src/config.rs:933]: Please use `testutils::new_temp_dir()` to create tempdir with predictable name.

File: src/config.rs [hunk 5]
```diff
@@ -632,4 +732,286 @@ mod tests {
...
+    impl TestCase {
+        fn config(&self, root: &Path) -> ConfigEnv {
+            ConfigEnv {
+                config_dir: self.cfg.config_dir.as_ref().map(|p| root.join(p)),
+                home_dir: self.cfg.home_dir.as_ref().map(|p| root.join(p)),
+                jj_config: self
+                    .cfg
+                    .jj_config
+                    .as_ref()
+                    .map(|p| root.join(p).to_string_lossy().to_string()),
```
Comment (yuja) [inline src/config.rs:916]: nit: better to panic instead of using lossy value in tests

File: src/config.rs [hunk 6]
```diff
@@ -632,4 +732,286 @@ mod tests {
...
+            files: vec!["home/.jjconfig.toml"],
+            cfg: ConfigEnv {
+                home_dir: Some("home".into()),
+                ..Default::default()
+            },
+            want: Want::ExistingAndNew("home/.jjconfig.toml"),
+            ..Default::default()
+        }
+        .run()
+    }
```
Comment (yuja) [inline src/config.rs:748]: Might be better to use some parameterized testing library than crafting micro framework. We use `test-case` in lib crate, but I'm not sure if it's convenient here.

@martinvonz do you have any recommendation?

Comment (martinvonz) [reply src/config.rs:748]: Yes, `test-case` might work. I considered switching from `test-case` to `rstest` but I don't know if it's really better in any way we care about (it supports fixtures, but I don't know if that would help us anyway).

Comment (qfel) [reply src/config.rs:748]: Hm it doesn't seem to work great for nontrivial case descriptions. I *can* use test-case, basically something like

```
    #[test_case(
        TestCase {
            files: vec!["home/.jjconfig.toml", "config/jj/config.toml"],
            cfg: ConfigEnv {
                home_dir: Some("home".into()),
                config_dir: Some("config".into()),
                ..Default::default()
            },
            want: Want::Fail(|err| match err {
                ConfigError::AmbiguousSource(_, _) => Ok(()),
                _ => Err(anyhow::anyhow!(
                    "expected AmbiguousSource error, got: {err}"
                )),
            }),
            ..Default::default()
        }; "path_ambiguous")]
    #[test_case(
        TestCase {
            cfg: Default::default(),
            want: Want::None,
            ..Default::default()
        }; "path_none")]
    fn config_path(case: TestCase) -> anyhow::Result<()> {
        case.run()
    }
```

but the case contents will not be autoformatted due to the macro magic. Is it the direction you'd rather go? I do have to say I miss Go's framework.

Comment (yuja) [reply src/config.rs:748]: Maybe we can extract factory functions?
```
fn foo() -> TestCase { .. }
#[test_case(foo())]
#[test_case(bar())]
...
```
Not tried with `test_case`, but I think `rstest` does support this.

Comment (qfel) [reply src/config.rs:748]: How is that better from the current version, where there is a separate function for each case?

Comment (yuja) [reply src/config.rs:748]: I think we can inline `TestCase::run()` and rewrite/split to make it more easier to follow. I don't know if it's better to serve all parameters with a single test function, though.

And if each `test_case` is labeled, maybe we can remove some naming wrappers like `subrun()`?

Comment (qfel) [reply src/config.rs:748]: > I think we can inline TestCase::run() and rewrite/split to make it more easier to follow. 

If we choose to use `test_case`, sure. But to be clear, you do prefer `test_case` over separate tests (`cargo fmt` works only with the latter)?

> I don't know if it's better to serve all parameters with a single test function, though.

I suspect having a few independent test case formats would result in more code duplication/boilerplate. Also with one test case format it's easier to understand what is being tested, since you have to understand only one format / it's more consistent.

`subrun()` is there to distinguish between `new_config_path` and `existing_config_path` - which are exercices by each test case. So it's not redundant with the test case name.

Comment (yuja) [reply src/config.rs:748]: I prefer simple (and dumb) tests like
```rust
    #[test]
    fn test_config_path_none() {
        let config_env = set_up_files(...);
        ...
        assert_eq!(config_env.existing_config_path(), ...);
        assert_eq!(config_env.new_config_path(), ...);
        ...
    }
```

If `test_case()` helps to deduplicate test code (without sacrificing readability), that's nice.

My feeling is that the current `TestCase::run()` is too structured enough to write a test for the test runner itself.

> you do prefer test_case over separate tests (`cargo fmt` works only with the latter)?

If you mean embedding large code block in `#[test_case(...)]`, no, I wouldn't prefer that.

File: src/config.rs [hunk 7]
```diff
@@ -190,28 +192,126 @@ impl LayeredConfigs {
...
+            Some(path) if path.exists() => ConfigPath::Existing(path),
+            Some(path) => ConfigPath::New(path),
+            None => ConfigPath::Unavailable,
+        }
+    }
+}
+
+fn create_config_file(path: &Path) -> std::io::Result<std::fs::File> {
+    if let Some(parent) = path.parent() {
+        std::fs::create_dir_all(parent)?;
```
[the remark this answers is not in the document]
Comment (necauqua) [reply src/config.rs:217]: Well something is very wrong if `~` does not exist, and just ensuring the abstract 'config folder' (which is XDG_CONFIG on linux) exists is fine I think

[the remark this answers is not in the document]
Comment (yuja) [reply src/config.rs:217]: I was wondering if `~/.config` should be set to `0700`. (For subdirectory, we wouldn't have to care so long as the parent directory isn't world readable.)

The spec says:
> If, when attempting to write a file, the destination directory is non-existent an attempt should be made to create it with permission `0700`. If the destination directory exists already the permissions should not be changed.
https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

So I was wrong. We should create both `~/.config` and `~/.config/jj` with mode `0700`? Apparently, there are many programs that creates subdirectory with mode `~umask`, though.

File: src/config.rs (hunk 3 above)
Comment (qfel) [reply src/config.rs:255]: Was thinking about it before, but the impl got a bit messy - though potentially because I wanted to fallback to using one directory if another failed to create. Will circle back here once we get clarity on the directory creation.

File: src/config.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (qfel) [reply src/config.rs:217]: So our options are:
 1. Create the entire path with create_dir_all (current, simple, questionable behavior)
 2. Do not create any directories (you didn't like that)
 3. Create directories but only on top of standard directories (ie. potentially create the jj subdirectory)
 4. Create directories but only on top of ~ (conformant to the XDG spec if we use the right mode, which also means no create_dir_all)

Should I do (4), or still needs more discussion?

File: src/config.rs (hunk 6 above)
Comment (qfel) [reply src/config.rs:745]: It had Default solely to make TestCase Default, but given the current code there is no longer benefit to Default TestCase, so removed.

File: CHANGELOG.md [hunk 8]
```diff
@@ -62,6 +62,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### New features
 
+* `jj config edit --user` and `jj config set --user` will now pick a default
+  config location if no existing file is found, potentially creating parent directories.
+
```
Comment (martinvonz) [inline CHANGELOG.md:67]: This should go under 0.8.0 :) (I assume it's here because we released 0.8.0 after you started working on this PR).

File: src/cli_util.rs [hunk 9]
```diff
@@ -2008,7 +2008,7 @@ pub fn get_config_file_path(
     let edit_path = match config_source {
         // TODO(#531): Special-case for editors that can't handle viewing directories?
         ConfigSource::User => {
-            config_path()?.ok_or_else(|| user_error("No repo config path found to edit"))?
+            new_config_path()?.ok_or_else(|| user_error("No repo config path found to edit"))?
```
Comment (yuja) [inline src/cli_util.rs:2011]: nit: let's rename `get_config_file_path()` to something like `get_new_config_file_path()` to prevent future misuse.

File: src/config.rs [hunk 10]
```diff
@@ -190,30 +192,126 @@ impl LayeredConfigs {
...
-pub fn config_path() -> Result<Option<PathBuf>, ConfigError> {
-    if let Ok(config_path) = env::var("JJ_CONFIG") {
-        // TODO: We should probably support colon-separated (std::env::split_paths)
-        // paths here
-        Ok(Some(PathBuf::from(config_path)))
-    } else {
+enum ConfigPath {
+    /// Path to an existing config file. Note that this is not necessarily
+    /// checked (eg. if specified via the JJ_CONFIG env var).
+    Existing(PathBuf),
```
Comment (yuja) [inline src/config.rs:198]: nit: I think `JJ_CONFIG` case also uses checked constructor.

**Review by yuja — APPROVED** [review]
This LGTM, thanks.

**Comment by martinvonz** [comment]
@qfel, thanks again for this PR! Feel free to merge it once you've addressed all feedback. It looks like there's just some remaining change requested by @necauqua. I don't even know how to tell what that change was, but if you know what it was and you've addressed it, I guess you can dismiss it.

**Review by necauqua — APPROVED** [review]
Oh yeah, github UI, right. My initial change request was to make it actually work with the xdg path because it's unlikely that the`jj` folder existed there :)
